addons/pos_gift_card/static/src/js/GiftCardPopup.js

```diff
@@ -37,17 +39,34 @@ odoo.define("pos_gift_card.GiftCardPopup", function (require) {
       this.state.showGiftCardDetails = true;
     }
 
-    addGiftCardProduct(giftCard) {
+    async addGiftCardProduct(giftCard) {
       let gift =
         this.env.pos.db.product_by_id[
           this.env.pos.config.gift_card_product_id[0]
         ];
-      this.env.pos.get_order().add_product(gift, {
-        price: this.state.amountToSet,
-        quantity: 1,
-        merge: false,
-        generated_gift_card_ids: giftCard ? giftCard.id : false,
-      });
+
+      let can_be_sold = true;
+      if (giftCard) {
+        can_be_sold = await this.rpc({
+          model: "gift.card",
+          method: "can_be_sold",
+          args: [giftCard.id],
+        });
+      }
+
+      if (can_be_sold) {
+        this.env.pos.get_order().add_product(gift, {
+          price: this.state.amountToSet,
+          quantity: 1,
+          merge: false,
+          generated_gift_card_ids: giftCard ? giftCard.id : false,
+        });
+      } else {
+        await this.showPopup('ErrorPopup', {
+          'title': _t('This gift card has already been sold'),
+          'body': _t('You cannot sell a gift card that has already been sold'),
+        });
+      }
     }
 
     async getGiftCard() {
```

Comment:
it seems to make a lot of change, and maybe we can ease it. Can't we just change this "GetGiftCard" function, to get all fields. And so after that we know that if there are buy lines and/or pos buy lines? We'll be able to know if it has already been sold in the js.

Reply:
I tried to do it this way the first time but it was not working. The getGiftCard method is only retrieving the gift card from the server if it's not already in the PoS. Wich is the case if the gift card was created before the opening of the PoS so the version of the gift card is not the right one. However, we could change the getGiftCard to update the gift card if it's already in the PoS. What do you think is the best solution?

Reply:
It is because of:
"      let giftCard = this.env.pos.giftCard.find(
        (gift) => gift.code === this.state.giftCardBarcode
      );
 "
in getGiftCard, which allow the use of gift cart "Offline", but as you want to check the validity on the server each tim, you can just remove it, remove the load of existing giftcards in models, and jus let the search_read, without fields to get all of them